addons/web/models/models.py

```diff
@@ -295,89 +298,115 @@ def search_panel_global_counters(self, values_range, parent_name):
+                        { 'id': id, parent_name: False or (id, display_name),... }
+        :param parent_name, string, indicates which key determines the parent
+        :param ids: list of record ids
+        :return: the sublist (as a filter object) of records with the above properties
+        }
+        """
+        def get_parent_id(record):
+            value = record[parent_name]
+            return value and value[0]
+
+        allowed_records = { record['id']: record for record in records }
+        records_to_keep = {}
+        for id in ids:
+            record_id = id
+            ancestor_chain = {}
+            chain_is_fully_included = True
+            while chain_is_fully_included and record_id:
+                known_status = records_to_keep.get(record_id)
+                if known_status != None:
+                    # the record and its known ancestors have already been considered
+                    chain_is_fully_included = known_status
+                    break
+                record = allowed_records.get(record_id)
+                if record:
+                    ancestor_chain[record_id] = record
+                    record_id = get_parent_id(record)
+                else:
+                    chain_is_fully_included = False
 
-        :param field_name: the name of a filter field;
-            possible types are many2one, many2many, selection.
-        :param category_domain: domain generated by categories
-                                (this parameter is used in search_panel_local_counters)
-        :param comodel_domain: domain of field values (if relational)
-                                (this parameter is used in search_panel_local_counters)
-        :param comodel_field_names: a dict with keys given by field names of the comodel
-                                    of field_name (if relational).
-                                    the values are functions used to process
-                                    the field names values. They must accept the single
-                                    parameter record (see below)
-        :param enable_counters: whether to count records by value. This is done in
-                                    search_panel_local_counters
-        :param filter_domain: domain generated by filters
-                                (this parameter is used in search_panel_local_counters)
-        :param search_domain: base domain of search
-                                (this parameter is used in search_panel_local_counters)
-        :return: a list of possible values, each being a dict with keys
-            'id' (value),
-            'name' (value label),
-            '__count' (how many records with that value),
-            and possibly other keys
+            for id, record in ancestor_chain.items():
+                records_to_keep[id] = chain_is_fully_included
+
+        # we keep initial order
+        return filter(lambda rec: records_to_keep.get(rec['id']), records)
```

Comment:
`filter` is lazy, use a  list comprehension instead:
```suggestion
        return [rec for rec in records if records_to_keep.get(rec['id'])]
```

Reply:
ok